source/agora/consensus/validation/Block.d

```diff
@@ -112,6 +115,151 @@ public string isInvalidReason (const ref Block block, in ulong prev_height,
+
+    UTXOSetValue[Hash] utxo_set;
+    foreach (const ref tx; genesisBlock.txs)
+    {
+        if (!(tx.type == TxType.Payment || tx.type == TxType.Freeze))
+            return "GenesisBlock: Transaction type allows only "
+                ~ "Payment and Freeze";
+
+        if (tx.inputs.length != 1)
+            return "GenesisBlock: Transaction must have only one input";
+
+        if (tx.outputs.length == 0)
+            return "GenesisBlock: No output(s) in the transaction";
+
+        Hash tx_hash = tx.hashFull();
+        foreach (idx, const ref output; tx.outputs)
+        {
+            if (output.address == PublicKey.init)
+                return "GenesisBlock: The address of the output(s) is empty"
+                    ~ "in the transaction";
+
+            // disallow negative amounts
+            if (!output.value.isValid())
+                return "GenesisBlock: Output(s) overflow or underflow"
+                    ~ "in the transaction";
+
+            // disallow 0 amount
+            if (output.value == Amount(0))
+                return "GenesisBlock: Value of output is 0"
+                    ~ "in the transaction";
+
+            if (tx.inputs[0].signature != Signature.init
+                && output.address.verify(tx.inputs[0].signature, tx_hash[]))
+                return "GenesisBlock: Input has invalid signature "
+                    ~ "in the transaction";
+
+            const utxo_value = UTXOSetValue(genesisBlock.header.height,
+                tx.type, output);
+            utxo_set[UTXOSet.getHash(tx_hash, idx)] = utxo_value;
+        }
+    }
+
+    Hash[] merkle_tree;
+    if (genesisBlock.header.merkle_root !=
+        Block.buildMerkleTree(genesisBlock.txs, merkle_tree))
+        return "GenesisBlock: Merkle root does not match header's";
+
+    Set!Hash used_utxos;
+
+    bool findUTXO (Hash hash, size_t index, out UTXOSetValue value)
+        nothrow @safe
+    {
+        Hash utxo_hash;
+
+        if (index == size_t.max)
+            utxo_hash = hash;
+        else
+            utxo_hash = UTXOSet.getHash(hash, index);
+
+        if (utxo_hash in used_utxos)
+            return false;  // double-spend
+
+        if (auto ptr = utxo_hash in utxo_set)
+        {
+            value = *ptr;
+            used_utxos.put(utxo_hash);
+            return true;
+        }
+        return false;
+    }
+
+    // If there are no enrollments, return them here early.
+    if (genesisBlock.header.enrollments.length == 0)
+        return null;
+
+    if (!isStrictlyMonotonic!"a.utxo_key < b.utxo_key"
+        (genesisBlock.header.enrollments))
+        return "GenesisBlock: The enrollments should be arranged in "
+            ~ "ascending order by the utxo_key";
+
+    foreach (const ref enrollment; genesisBlock.header.enrollments)
+    {
+        if (auto fail_reason = VEn.isInvalidReason(enrollment, &findUTXO))
+            return "GenesisBlock: " ~ fail_reason;
```

Comment:
Return `fail_reason`. Those routines should not allocate.

Reply:
I fixed it.